### DISCUSSION TITLE
Massive JSON text files

### DISCUSSION THREAD

**Discussion by Mauvila** [CMSgov/price-transparency-guide#528] [General]
Per Cigna's TIC machine readable webpage:
> Each of the files accessible through the Table of Contents is in JSON format and may be as large as one Terabyte (TB) in file size, which has significant system requirements for use. Please ensure you have the required memory capacity, hardware, and software capabilities before attempting to download.

I have 128 GB of RAM, but I don't feel my machine would be able to tackle parsing a terabyte JSON file.  These files make the 80 GB JSON text files on UHC's site seem small.   Were these file sizes expected when the rules were issued, or is this an attempt by payors to keep everything opaque?

What is the expected general strategy of analyzing these mega-files?  I was not aware that a single file could be this large, and the fact that it's JSON (ie tree like structure) makes it even more difficult, as you wouldn't simply be able to split it up by new lines as if it were a huge CSV file.  I imagine one would need a special tool to parse the files, upload them to a large cloud-based database (NoSQL?). 

If this is not the appropriate forum for this type of question, please point me in the right direction.

**Comment by ghost** [comment]
Welcome to the party and you must be referencing the In-Network Negotiated Rates file! 

The file size was a known issue based on the requirements of the ruling for the last 2 years. A payor is expected to provide a file that will contain **all possible**, not just claims billed, procedures/billing codes a provider can bill which can be referred to in the description of "_plan or issuer must disclose in-network **provider negotiated rates for all items and services**_". 

The current In-Network Negotiated Rates file would be even larger if it was not for the effort of this community in working with Scott of CMS to optimize the layout to take advantage of mapping duplicate data in arrays and Table of Contents.

**Reply by aaronmefford** [reply]
They could just as easily been JSONL files with minimal overhead, significantly reducing the complexity of parsing.

**Comment by Devyandu** [comment]
> 

The huge file sizes have been known for a while. If you are going to be consuming these files, you should look at stream based JSON parsers so that you do not load the whole JSON as a single object in memory.

See this : [https://howtodoinjava.com/gson/jsonreader-streaming-json-parser/](https://howtodoinjava.com/gson/jsonreader-streaming-json-parser/). A 128Gig RAM machine should be more than sufficient to parse these files if you use a streaming library.
It should be a fairly straightforward task to take the huge JSON and convert it to a (somewhat relational) set of 5-6 CSVs.

**Comment by escoses100** [comment]
If you just want to read them and the files fit on your laptop unzipped, you can use a hex editor like Hex Fiend that does not load the entire file.  There is no formatting, but search works ok (it has a progress bar) and you can get a look at how the file is generally formatted.

**Comment by shaselton-usds** [comment]
Just to echo what everyone on here has been saying, these requirements translate to quite a bit of information being released for the first time. This community was instrumental in driving improvements. I would suggest streaming parsers if you plan to work on the files locally (this is how the [schema validator works](https://github.com/CMSgov/price-transparency-guide-validator) -- basically impossible to validate GBs locally without streaming). If you're looking to just take a peek in the file, you can always take the first 100 or 1000 lines of the file to read. If you're looking to do a lookup without actually opening the file, `grep` is a great command.

At the end of the day, serious analysis of these files require them to be translated/transferred into a data warehouse for deep queries.

**Reply by sbenphia** [reply]
> Just to echo what everyone on here has been saying, these requirements translate to quite a bit of information being released for the first time. This community was instrumental in driving improvements. I would suggest streaming parsers if you plan to work on the files locally (this is how the [schema validator works](https://github.com/CMSgov/price-transparency-guide-validator) -- basically impossible to validate GBs locally without streaming). If you're looking to just take a peek in the file, you can always take the first 100 or 1000 lines of the file to read. If you're looking to do a lookup without actually opening the file, `grep` is a great command.
> 
> At the end of the day, serious analysis of these files require them to be translated/transferred into a data warehouse for deep queries.

I think this community may have missed the mark.  Fee schedules (the in-network files) like this traditionally do not require this level of work and have existed for years in machine readable formats in workers compensation, medicaid, personal injury protection, and payor, provider, and TPA platforms in much more manageable datasets, I think the first step in serious analysis will be to start pairing/distilling these terabyte files down to the information the law intends to expose - the slightly varied 25-30 page contracts (with about 15 megs of associated spreadsheets) that carriers negotiate every year and that providers then load into their billing systems. 

At the end of the day, serious analysis of these files will occur in excel once they are distilled back to the simple contract forms of their origin - if we can get them there (in my humble opinion).  That said - this is a start and I appreciate the fact there is now data to attempt to review and the conversation has started in the open (even though I had to buy a hard drive over the weekend).  Heck - even the fact this is all on github is a great accomplishment.

**Reply by aaronmefford** [reply]
unless the lines you need dont occur until 10000 lines into the file.

**Comment by dgolden** [comment]
@sbenphia - how would I go about obtaining an example of these, to compare against the resultant In-Network Rates MRF?
 
"25-30 page contracts (with about 15 megs of associated spreadsheets) that carriers negotiate every year and that providers then load into their billing systems."

**Reply by sbenphia** [reply]
Generally payor/provider contracts are confidential as private contracts with price terms - sometimes they pop up in court records when hospitals and carriers sue each other, but those are pretty heavily redacted and a little tough to find buried in the costly pacer docket (courtlistener might change that a little).  If you consult for either payors or providers on contracting (or represent them as an attorney) you'll see the contracts and schedules, but you'll have a duty of confidentiality.  Another place where you could review similar approaches and methodologies would be in the workers compensation and personal injury space, where the government essentially writes a public contract between providers and work comp payors in the form of a fee schedule, however, even many of those are behind a paywall and have use licenses because the contractor developing them charges for access (https://orders.fairhealth.org/).   One other place you might get these (if they don't find an exception) is through public records requests where the hospital/provider is a public (or quasi-public) entity and the other party is not.
